[Guide]: Instant navigation experiences (#3300)
GoogleChrome/web.dev: src/site/content/en/reliable/instant-navigation-experiences/index.md

To speed up these requests, if you can anticipate the user's action, you can request these pages and assets beforehand and keep them in the cache for a short period of time until the user clicks on these links. This technique is called [prefetching](https://web.dev/link-prefetch/) and it's commonly implemented by adding `<link rel="prefetch">` tags to pages, indicating the resource to prefetch.

In this guide we'll explore different ways in which [service workers](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) can be used as a complement of traditional prefetching techniques.

## Production cases

[MercadoLibre](https://www.mercadolibre.com.ar/) is the biggest e-commerce site in Latin America. To speed up navigations, they dynamically inject `<link rel=”prefetch”>` tags in some parts of the flow. For example, in listing pages, they fetch the next result page as soon as the user scrolls to the bottom of the listing:

<figure class="w-figure">
  <img src="mercadolibre-prefetch.png" 
       alt="Screenshot of MercadoLibre's listing pages onea and two and a Link Prefetch tag connecting both.">
</figure>

> Prefetched files are requested at the "Lowest" priority and stored in the [HTTP cache](https://web.dev/http-cache/) or the [memory cache](https://calendar.perfplanet.com/2016/a-tale-of-four-caches/) (depending on whether the resource is cacheable or not), for an amount of time that varies by browsers. For example, as of Chrome 56, this value is 5 minutes. Resources are kept around for five minutes, after which the normal `Cache-Control` rules for the resource apply.

Reviewer: ```suggestion
Prefetched files are requested at the "Lowest" priority and stored in the [HTTP cache](https://web.dev/http-cache/) or the [memory cache](https://calendar.perfplanet.com/2016/a-tale-of-four-caches/) (depending on whether the resource is cacheable or not), for an amount of time that varies by browsers. For example, as of Chrome 85, this value is 5 minutes. Resources are kept around for five minutes, after which the normal `Cache-Control` rules for the resource apply.
```
I know I asked you to specify when this become true, but I'm realizing now we should actually mention the current version of Chrome in which this applies, because talking about Chrome 56 makes this post seem ancient

{% endAside %}

### 1. Precache static pages and page subresources

[Precaching](https://web.dev/precache-with-workbox/) is the ability of the service worker to save files to the cache while it's installing. 

{% Aside %}
Precaching sounds similar to prefetching, but it's a different technique. In the first one, the service worker fetches and stores resources (typically static files) while it's installing and keeps them in the cache until a new version of the file is available. In the second, resources are requested ahead of time to have it in the cache for brief periods of time in order to speed up subsequent navigations. 

> {% endAside %}

Reviewer: This is great, thank you

> In the following cases precaching is used to achieve similar goal than prefetching: making navigations faster.

Reviewer: ```suggestion
In the following cases precaching is used to achieve a similar goal as prefetching: making navigations faster.
```

```javascript
workbox.precaching.precacheAndRoute([
  {url: '/about.html', revision: 'abcd1234'},
  // ... other entries ...
]);
```

#### Precaching page subresources 

Precaching static assets that the different sections of the site might use (e.g. JavaScript, CSS, etc.), is a general best practice and can give an extra boost in prefetching scenarios.

To speed up navigations in an e-commerce site, you can use  `<link rel="prefetch">` tags in listing pages to prefetch product detail pages for the first few products of a listing page. If you have already precached the product page subresources, this can make the navigation even faster.

To implement this:

- Add a `<link rel="prefetch">` tag to the page:

```html
 <link rel="prefetch" href="/phones/smartphone-5x.html" as="document">
```

- Add the page subresources to the precache list in the service worker:

```javascript
workbox.precaching.precacheAndRoute([
  '/styles/product-page.ac29.css',
  // ... other entries ...
]);
```

### 2. Extend the lifetime of prefetch resources

> As mentioned earlier, `<link rel="prefetch">` fetches and keeps resources in the HTTP cache for a limited amount of time, after which point the `Cache-Control` rules for a resource apply. As of Chrome 56, this value is 5 minutes.

Reviewer: ```suggestion
As mentioned earlier, `<link rel="prefetch">` fetches and keeps resources in the HTTP cache for a limited amount of time, after which point the `Cache-Control` rules for a resource apply. As of Chrome 85, this value is 5 minutes.
```

Reviewer: I'm just thinking about this now: do we have any knowledge about how the other browsers behave? That would be nice to include. This isn't a blocker but if you do know this information, it would be nice to include.

Author: Hi Kayce, this is an important point. As this is not part of the spec, each browser can implement the prefetching window differently, and there's no public information about it. At the time when we wrote https://web.dev/link-prefetch/ (the canonical guide), we were able to get this info from Chrome folks, but unfortunately couldn't find details about the rest.
